crates/gitbutler-tauri/src/workspace.rs

```diff
@@ -29,3 +37,160 @@ pub fn stack_branches(
+#[tauri::command(async)]
+#[instrument(skip(projects), err(Debug))]
+pub fn hunk_dependencies_for_workspace_changes(
+    projects: State<'_, projects::Controller>,
+    project_id: ProjectId,
+) -> Result<HunkDependencies, Error> {
+    let project = projects.get(project_id)?;
+    let dependencies =
+        hunk_dependencies_for_workspace_changes_by_worktree_dir(&project.path, &project.gb_dir())?;
+    Ok(dependencies)
+}
+
+pub fn hunk_dependencies_for_workspace_changes_by_worktree_dir(
+    worktree_dir: &Path,
+    gitbutler_dir: &Path,
+) -> anyhow::Result<HunkDependencies> {
+    let repo = gix::open(worktree_dir).map_err(anyhow::Error::from)?;
+    let worktree_changes = but_core::diff::worktree_changes(&repo)?;
+    let stacks = but_workspace::stacks(gitbutler_dir)?;
+    let common_merge_base = gitbutler_stack::VirtualBranchesHandle::new(gitbutler_dir)
+        .get_default_target()?
+        .sha;
+    let input_stacks = but_hunk_dependency::workspace_stacks_to_input_stacks(
+        &repo,
+        &stacks,
+        common_merge_base.to_gix(),
+    )?;
+    let ranges = but_hunk_dependency::WorkspaceRanges::try_from_stacks(input_stacks)?;
+    HunkDependencies::try_from_workspace_ranges(&repo, ranges, worktree_changes.changes)
+}
+
+/// Calculate as hash for a `universal_diff`.
+// TODO: see if this should be avoided entirely here as the current impl would allow for hash collisions.
+pub fn hash_lines(universal_diff: impl AsRef<[u8]>) -> HunkHash {
+    let diff = universal_diff.as_ref();
+    assert!(
+        diff.starts_with(b"@@"),
+        "BUG: input mut be a universal diff"
+    );
+    let mut ctx = rustc_hash::FxHasher::default();
+    diff.lines_with_terminator()
+        .skip(1) // skip the first line which is the diff header.
+        .for_each(|line| ctx.write(line));
+    ctx.finish()
+}
+
+/// A way to represent all hunk dependencies that would make it possible to know what can be applied, and were.
+///
+/// Note that the [`errors`](Self::errors) field may contain information about specific failures, while other paths
+/// may have succeeded computing.
+#[derive(Debug, Clone, Serialize)]
+pub struct HunkDependencies {
+    /// A map from diffs to branch and commit dependencies.
+    // TODO: could this be a specific type? Is the mapping truly required?
+    //       Is this because `commit_dependent_diffs` use `HunkHash`?
+    // TODO: the frontend actually has no way of associating the hunks it gets with this hash as it's made
+    //       on the patch lines without any context lines, while it has context lines.
+    //       Hash must then skip the context lines if there are any.
+    pub diffs: Vec<(HunkHash, Vec<HunkLock>)>,
+    /// A map from stack id to commit dependencies.
+    /// Commit dependencies map commit id to commits it depends on.
+    // TODO: have to use strings as keys for serialization (`gix::ObjectId`)
+    #[serde(skip)]
```

Comment:
Something that comes up to my mind while reading this - a lot of this data from the v2 flow was added "because we could" but it was never used by the frontend. 
For example the inverse dependencies, we might not need to calculate / present to the fronted.

Perhaps in the spirit of keeping it simple, we could start off with just the plain hunk dependency (currently in the diffs field) and nothing else, if that makes things easier.

Reply:
Thanks, your comment helped me to go all in on the culling, which helped to remove a lot of code and plenty of tests which now would show no changes at all. Or in other words, the tests were using commit-dependencies as a way to understand about hunk relationships.
This, however, is already 'pinned' in `but-hunk-dependency` using `insta`.
My thinking is that tests can be re-added once it's clear what's needed.